cpp/velox/shuffle/VeloxShuffleReader.cc

```diff
@@ -321,6 +321,10 @@ VeloxShuffleReader::VeloxShuffleReader(
 arrow::Result<std::shared_ptr<ColumnarBatch>> VeloxShuffleReader::next() {
   ARROW_ASSIGN_OR_RAISE(auto batch, Reader::next());
   if (batch == nullptr) {
+    // Some empty batch has not been released here, so check is needed.
+    if (veloxPool_->currentBytes() == 0 && veloxPool_->reservedBytes() == 0) {
+      veloxPool_->shrinkManaged(veloxPool_.get(), veloxPool_->capacity());
+    }
```

Comment:
veloxPool_ can't be reused by next shuffle reader, right? So the number of used memory keeps increasing.

We may avoid explicit calls to `shrinkManaged` since it's not a public/stable Velox API. Would you like to help find out if we can reuse the pool here, probably? Thanks!

Reply:
> veloxPool_ can't be reused by next shuffle reader, right?

Yes, each shuffle reader stream has its pool, and each `ShuffleReaderJniWrapper#make` call will create a new pool. 

> Would you like to help find out if we can reuse the pool here, probably?

That's a better way for me too, and it may bring lots of refactoring, current way is just a hot fix for now, if you accept this idea.

I actually want to a whole refactoring/optimization as I discussed with you, but I need learn both Gluten and Velox codes more deeply and will keep working on it.

Reply:
And I think reuse is not the key, each memory consumer should support spill is the key, and some spill operations just shrink itself.

Reply:
> And I think reuse is not the key, each memory consumer should support spill is the key

The Velox pools should be all give their reservations back to Spark once destructed. Currently shuffle reader is a anti-case since all the pools are destructed at the point that task ends, that's too late. We actually have several solutions:

option 1: This patch
option 2: Reuse shuffle reader's memory pool
option 3: Close shuffle reader early, before task ends
option 4: Let memory pool reports it's used bytes to Spark, not capacity

I think we can first implement 1, maybe 2 and 3 too, to make this whole shuffle reader stuff more robust that it is.

Reply:
op3 is similar with op1, but check of currentBytes and TaskResources.addRecycle are still needed, cause I found some empty batch will exist when current shuffle read stream close, we need consider more careful way for batch later.

op4 seems same with origin design which make spark consumer bind with allocator, now spark consumer bind with arbitrator. I think bind with reservation/capacity changes is better.

Reply:
Thanks for the explanation.

Let's apply this patch as a temporary solution, since the API `shrinkManaged` will not last for long time there for gluten.

Once op2 is implemented, we can drop this part of code. What do you think about it?

Reply:
> since the API shrinkManaged will not last for long time there for gluten.

Why and how do you shrink when nativeSpill be called if this API be removed?

> Once op2 is implemented, we can drop this part of code. What do you think about it?

LGTM.